File: src/main/java/pl/dmichalski/algorithms/_2_count_chars_in_text/CharCounterService2.java
package pl.dmichalski.algorithms._2_count_chars_in_text;

import java.util.HashMap;
import java.util.Map;

public class CharCounterService2 {

    /**
     * O(n) complexity
     */
    public Map<Character, Integer> charCount(String text) {
        Map<Character, Integer> charactersCounter = new HashMap<>();

        String lowerCaseText = text.toLowerCase();

        for (int i = 0; i < lowerCaseText.length(); i++) {
            if (isAlphanumeric(lowerCaseText.charAt(i))) {
                Character character = lowerCaseText.charAt(i);
                charactersCounter.putIfAbsent(character, 0);
                int increasedValue = charactersCounter.get(character) + 1;
                charactersCounter.put(character, increasedValue);
            }
        }

        return charactersCounter;
    }

    private boolean isAlphanumeric(char character) {
        return ((int) character > 47 && (int) character < 58)
                || ((int) character > 64 && (int) character < 91)
                || ((int) character > 96 && (int) character < 123);
    }

}

File: src/main/kotlin/pl/dmichalski/algorithms/algorithms/_2_count_chars_in_text/CharCounterService2.kt
package pl.dmichalski.algorithms.algorithms._2_count_chars_in_text

internal class CharCounterService2 {

    /**
     * O(n) time complexity
     */
    fun charCount(text: String): Map<Char, Int> {
        val charactersCounter: MutableMap<Char, Int> = HashMap()
        val lowerCaseText = text.toLowerCase()
        for (i in lowerCaseText.indices) {
            if (isAlphanumeric(lowerCaseText[i])) {
                val character = lowerCaseText[i]
                charactersCounter.putIfAbsent(character, 0)
                val increasedValue = charactersCounter[character]!! + 1
                charactersCounter[character] = increasedValue
            }
        }
        return charactersCounter
    }

    private fun isAlphanumeric(character: Char): Boolean {
        return (character.toInt() in 48..57
                || character.toInt() in 65..90
                || character.toInt() in 97..122)
    }
}
